<a href="https://colab.research.google.com/github/meenakshi930/DiseaseLeafClassifier/blob/anshika-data/Anshika/notebooks/04_data_split.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Splitting

## Disease Leaf Classifier

This notebook performs the final dataset splitting for the Disease Leaf Classifier project.

### Objective

The cleaned dataset is divided into three subsets using **stratified sampling**:

- **Training set:** 80%
- **Validation set:** 10%
- **Testing set:** 10%

Stratification is applied using the `class_name` column to maintain a similar distribution of disease classes across all three subsets.

### Input Dataset

The cleaned dataset contains **20,624 valid, non-duplicate images** belonging to **15 disease/healthy classes**.

The cleaned dataset is loaded from:

`data/clean_image_list.csv`

### Output Files

The following split files are generated:

- `data/train.csv`
- `data/validation.csv`
- `data/test.csv`

These files will be used for model training, validation, and final evaluation.

In [1]:
import pandas as pd

csv_url = "https://raw.githubusercontent.com/meenakshi930/DiseaseLeafClassifier/anshika-data/data/clean_image_list.csv"

clean_df = pd.read_csv(csv_url)

print("Total images:", len(clean_df))
print("\nColumns:")
print(clean_df.columns.tolist())

print("\nFirst 5 rows:")
print(clean_df.head())

Total images: 20624

Columns:
['class_name', 'image_name']

First 5 rows:
                      class_name  \
0  Pepper__bell___Bacterial_spot   
1  Pepper__bell___Bacterial_spot   
2  Pepper__bell___Bacterial_spot   
3  Pepper__bell___Bacterial_spot   
4  Pepper__bell___Bacterial_spot   

                                          image_name  
0  0f72db9c-d635-4415-9781-68937328aed2___NREC_B....  
1  1b8d3e98-43d9-441d-93ef-a359e6e9ddc2___NREC_B....  
2  b6d35d21-4812-4e32-ab9a-b6ddda8cbb79___JR_B.Sp...  
3  e4e7e5a2-5aa1-4378-833d-9142358c7c9b___JR_B.Sp...  
4  8b50f2db-cbf2-4758-bea6-a011f0cc1497___JR_B.Sp...  


In [2]:
from sklearn.model_selection import train_test_split

train_df, temp_df = train_test_split(
    clean_df,
    test_size=0.20,
    stratify=clean_df["class_name"],
    random_state=42
)

print("Training images:", len(train_df))
print("Temporary images:", len(temp_df))

Training images: 16499
Temporary images: 4125


In [3]:
val_df, test_df = train_test_split(
    temp_df,
    test_size=0.50,
    stratify=temp_df["class_name"],
    random_state=42
)

print("Training images:", len(train_df))
print("Validation images:", len(val_df))
print("Testing images:", len(test_df))
print("Total images:", len(train_df) + len(val_df) + len(test_df))

Training images: 16499
Validation images: 2062
Testing images: 2063
Total images: 20624


In [4]:
print("Training class distribution:")
print(train_df["class_name"].value_counts())

print("\nValidation class distribution:")
print(val_df["class_name"].value_counts())

print("\nTesting class distribution:")
print(test_df["class_name"].value_counts())

Training class distribution:
class_name
Tomato__Tomato_YellowLeaf__Curl_Virus          2566
Tomato_Bacterial_spot                          1701
Tomato_Late_blight                             1521
Tomato_Septoria_leaf_spot                      1417
Tomato_Spider_mites_Two_spotted_spider_mite    1341
Tomato_healthy                                 1268
Pepper__bell___healthy                         1182
Tomato__Target_Spot                            1123
Tomato_Early_blight                             800
Potato___Late_blight                            800
Potato___Early_blight                           800
Pepper__bell___Bacterial_spot                   798
Tomato_Leaf_Mold                                762
Tomato__Tomato_mosaic_virus                     298
Potato___healthy                                122
Name: count, dtype: int64

Validation class distribution:
class_name
Tomato__Tomato_YellowLeaf__Curl_Virus          321
Tomato_Bacterial_spot                          213
Tomato_La

In [5]:
import os

os.makedirs("data", exist_ok=True)

train_df.to_csv("data/train.csv", index=False)
val_df.to_csv("data/validation.csv", index=False)
test_df.to_csv("data/test.csv", index=False)

print("Split files saved successfully!")
print("\nFiles created:")
print("data/train.csv")
print("data/validation.csv")
print("data/test.csv")

Split files saved successfully!

Files created:
data/train.csv
data/validation.csv
data/test.csv


In [9]:
import os

for file in ["data/train.csv", "data/validation.csv", "data/test.csv"]:
    print(file, "→", os.path.exists(file))

data/train.csv → True
data/validation.csv → True
data/test.csv → True


In [11]:
from google.colab import files

files.download("data/train.csv")
files.download("data/validation.csv")
files.download("data/test.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Data Split Summary

The cleaned dataset was divided using stratified sampling to preserve the class distribution across all subsets.

- Total clean images: 20,624
- Training set: 16,499 images (80%)
- Validation set: 2,062 images (10%)
- Test set: 2,063 images (10%)
- Number of classes: 15
- Random state: 42
- Stratification: Applied using `class_name`

The final split files were saved as:
- `data/train.csv`
- `data/validation.csv`
- `data/test.csv`

The split was performed after invalid and duplicate files were removed, reducing the risk of duplicate-image leakage between training, validation, and test sets.